packages/echo-core/src/actions/coreActions/extensions.ts

```diff
@@ -0,0 +1,108 @@
+import { getCoreContext } from '../../state/globalState';
+import { UnRegisterExtension } from '../../types/api';
+import { GlobalState } from '../../types/state';
+import { ExtensionRegistration, ExtensionRegistry } from './../../types/registry/extension.types';
+import { dispatch } from './globalActions';
+
+/**
+ * Core Action for registering an extension.
+ * Extensions always tie to a specific component in echopedia: with these, it is possible
+ * to augment and extend pre-defined, existing components in the echopediaWeb main app.
+ * @param extension The extension to register;
+ * @param {string} extension.key  Unique key to identify the extension.
+ * @param {string} extension.extends The name of the component that would use this extension.
+ * @param {React.FC} extension.component The React component to be used by the extended component
+ * @param {callback} extension.isVisible May be used by the extended component: if the given extension should be rendered or not.
+ * @param {Record<string, unknown>} extension.options May be used by the extended component: additional, custom options to use,
+ * if needed by the extendable component
+ * @returns {UnRegisterExtension} A function to clear the given registration from the core state.
+ */
+export function registerExtension(extension: ExtensionRegistration): UnRegisterExtension {
+    const extendableComponentName = extension.extends;
+    dispatch(getCoreContext(), (state: GlobalState) => {
+        const updatedExtensionArray = createUpdatedExtensionArray(
+            extension,
+            state.registry.extensions,
+            extendableComponentName
+        );
+
+        return {
+            ...state,
+            registry: {
+                ...state.registry,
+                extensions: {
+                    ...state.registry.extensions,
+                    [extendableComponentName]: updatedExtensionArray
+                }
+            }
+        };
+    });
+    return (): void => {
+        unRegisterExtension(extendableComponentName, extension.key);
+    };
+}
+
+function createUpdatedExtensionArray(
+    extensionToRegister: ExtensionRegistration,
+    extensionRegistry: ExtensionRegistry,
+    extendableComponentName: string
+): ExtensionRegistration[] {
+    const originalExtensionsArray = extensionRegistry[extendableComponentName] || [];
+    let updatedExtensionsArray: ExtensionRegistration[] = originalExtensionsArray;
+    const newKeyAlreadyExists = originalExtensionsArray.map((item) => item.key).includes(extensionToRegister.key);
+
+    if (newKeyAlreadyExists) {
+        console.error(
```

Comment:
I'm wondering if it's better to throw an exeption, so the bug is reported to us? Now we'll never know if this happens for a user, and a user doesn't read the log usually.
Or will throwing an error crash the whole page?

Reply:
My goal here would be 'let the spice flow' even if someone made a mistake by using an already existing key.
Yep, an unhandled error would crush the page: showing the error page in echopediaWeb.

I could throw an error and handled it here with handleError, so it get's to our log. Will look into it.

Reply:
As discussed: will throw an unhandled error here, to let the caller of this function handle the error.